### PR TITLE
Add Sprints Feature

### PR DESCRIPTION
**Pull request by GhiocelAndrei**
Sprints Feature Implementation. Solves Issue #26

### CODE REVIEW

File: IssueTrackerApplication/Services/AutoMapperProfile.cs [hunk 1]
```diff
@@ -48,6 +48,21 @@ public AutoMapperProfile()
             CreateMap<UserUpdatingDto, User>();
 
             CreateMap<UserLoginDto, LoginUserCommand>();
+
+            // Sprint Mapping
+            CreateMap<Sprint, SprintDto>();
+
+            CreateMap<Sprint, CreateSprintCommand>();
```
Comment (cristi-paraschiv) [inline IssueTrackerApplication/Services/AutoMapperProfile.cs:55]: why do you need this mapping? are you creating commands from entities?

File: IssueTrackerApplication/Services/IssuesService.cs [hunk 2]
```diff
@@ -37,5 +39,40 @@ public override async Task<Issue> CreateAsync(CreateIssueCommand entity, Cancell
 
             return await base.CreateAsync(entity, ct);
         }
+
+        public async Task AssignSprintToIssuesAsync(List<long> ids, long sprintId, CancellationToken ct)
+        {
+            try
+            {
+                using (TransactionScope transactionScope = new TransactionScope(TransactionScopeAsyncFlowOption.Enabled))
```
Comment (cristi-paraschiv) [inline IssueTrackerApplication/Services/IssuesService.cs:47]: `using var...` and lose the curly braces

File: IssueTrackerApplication/Services/IssuesService.cs [hunk 3]
```diff
@@ -37,5 +39,40 @@ public override async Task<Issue> CreateAsync(CreateIssueCommand entity, Cancell
...
+                {
+
+                    var issuesModified = await DbContext.Issues
+                                        .Where(issue => ids.Contains(issue.Id)).ToListAsync();
```
Comment (cristi-paraschiv) [inline IssueTrackerApplication/Services/IssuesService.cs:51]: pass `ct` to `ToListAsync`

File: IssueTrackerDataAccess/Migrations/AddNewColumnsToIssuesAndSprints.cs [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+{
+    [Migration(202307281124)]
+    public class AddNewColumnsToIssuesAndSprints : Migration
+    {
+        public override void Up()
+        {
+            Create.Column("SprintId")
+               .OnTable("Issues").AsInt64().Nullable();
+
+            Create.Column("Active")
```
[the remark this answers is not in the document]
Comment (GhiocelAndrei) [reply IssueTrackerDataAccess/Migrations/AddNewColumnsToIssuesAndSprints.cs:13]: I created the Sprints Table using the properties in this issue : https://github.com/GhiocelAndrei/IssueTrackerAPI/issues/26 and then I added the Active property so I had to update the table.

File: IssueTracker.Abstractions/Models/Sprint.cs [hunk 5]
```diff
@@ -0,0 +1,14 @@
+﻿namespace IssueTracker.Abstractions.Models
+{
+    public class Sprint : ICreationTracking, IModificationTracking
```
[the remark this answers is not in the document]
Comment (GhiocelAndrei) [reply IssueTracker.Abstractions/Models/Sprint.cs]: Solved in the latest commit.

File: IssueTrackerAPI/Controllers/SprintsController.cs [hunk 6]
```diff
@@ -0,0 +1,99 @@
...
+
+            return Ok();
+        }
+
+        // DELETE: api/Sprints/5
+        [HttpDelete("{id}")]
+        [OAuth(Scopes.SprintsWrite)]
+        public async Task<IActionResult> DeleteSprint(long id, CancellationToken ct)
+        {
+            await _sprintsService.DeleteAsync(id, ct);
```
[the remark this answers is not in the document]
Comment (GhiocelAndrei) [reply IssueTrackerAPI/Controllers/SprintsController.cs:85]: Should I also delete all the issues assigned to that sprint ?

File: IssueTrackerApplication/Services/AutoMapperProfile.cs (hunk 1 above)
Comment (GhiocelAndrei) [reply IssueTrackerApplication/Services/AutoMapperProfile.cs]: Deleted unnecessary mappings in the latest commit.

File: IssueTrackerApplication/Services/IssuesService.cs (hunk 2 above)
Comment (GhiocelAndrei) [reply IssueTrackerApplication/Services/IssuesService.cs]: Solved in the latest commit.

File: IssueTrackerApplication/Services/IssuesService.cs (hunk 3 above)
Comment (GhiocelAndrei) [reply IssueTrackerApplication/Services/IssuesService.cs]: Solved in the latest commit.

File: IssueTrackerApplication/Services/IssuesService.cs [hunk 7]
```diff
@@ -37,5 +39,40 @@ public override async Task<Issue> CreateAsync(CreateIssueCommand entity, Cancell
...
+
+                    if(issuesModified.Count < ids.Count)
+                    {
+                        throw new InvalidInputException("There is existing Issue with one of the provided Ids.");
```
[the remark this answers is not in the document]
Comment (GhiocelAndrei) [reply IssueTrackerApplication/Services/IssuesService.cs]: Solved in the latest commit.

File: IssueTrackerApplication/Services/SprintsService.cs [hunk 8]
```diff
@@ -0,0 +1,49 @@
...
+                using (TransactionScope transactionScope = new TransactionScope(TransactionScopeAsyncFlowOption.Enabled))
+                {
+                    var sprint = await base.CreateAsync(sprintDto.SprintDto, ct);
+
+                    await _issueService.AssignSprintToIssuesAsync(sprintDto.Ids, sprint.Id, ct);
+
+                    transactionScope.Complete();
+                }
+            }
+            catch (ThreadAbortException)
```
[the remark this answers is not in the document]
Comment (GhiocelAndrei) [reply IssueTrackerApplication/Services/SprintsService.cs]: I needed that exception before using "TransactionScopeAsyncFlowOption.Enabled" as an option to TransactionScope and I forgot to remove it

File: IssueTrackerAPI/Controllers/SprintsController.cs [hunk 9]
```diff
@@ -0,0 +1,99 @@
...
+        // POST: api/Sprints
+        [HttpPost]
+        [OAuth(Scopes.SprintsWrite)]
+        public async Task<ActionResult<Sprint>> PostSprint(SprintCreatingDto sprintDto, CancellationToken ct)
+        {
+            var sprintCommand = _mapper.Map<CreateSprintCommand>(sprintDto);
+
+            var createdSprint = await _sprintsService.CreateAsync(sprintCommand, ct);
+
+            return CreatedAtAction("PostSprint", new { id = createdSprint.Id }, createdSprint);
```
[the remark this answers is not in the document]
Comment (GhiocelAndrei) [reply IssueTrackerAPI/Controllers/SprintsController.cs:65]: I'm not sure what should I modify ? This is a POST request not a GET request

File: IssueTrackerAPI/Controllers/IssuesController.cs [hunk 10]
```diff
@@ -53,6 +63,17 @@ public async Task<IActionResult> PutIssue(long id, IssueUpdatingDto issueDto, Ca
             return NoContent();
         }
 
+        // PUT: api/Issues/5
+        // To protect from overposting attacks, see https://go.microsoft.com/fwlink/?linkid=2123754
```
[the remark this answers is not in the document]
Comment (GhiocelAndrei) [reply IssueTrackerAPI/Controllers/IssuesController.cs]: Deleted links in the latest commit, they were generated by Visual Studio.

File: IssueTrackerApplication/Services/SprintsService.cs [hunk 11]
```diff
@@ -0,0 +1,49 @@
...
+        public async Task CloseSprint(long id, CancellationToken ct)
+        {
+            var sprint = await base.GetAsync(id, ct);
+
+            if (sprint == null)
+                throw new NotFoundException("The requested entity could not be found.");
+
+            sprint.Active = false;
+
+            await DbContext.SaveChangesAsync();
```
[the remark this answers is not in the document]
Comment (GhiocelAndrei) [reply IssueTrackerApplication/Services/SprintsService.cs]: Solved in the latest commit.

File: IssueTrackerAPI/Controllers/SprintsController.cs (hunk 9 above)
[the remark this answers is not in the document]
Comment (cristi-paraschiv) [reply IssueTrackerAPI/Controllers/SprintsController.cs:65]: this action returns via the Location header the URL where this resource is available
https://www.rfc-editor.org/rfc/rfc9110#status.201

see other usages of `CreatedAtAction` in the project

File: IssueTrackerAPI/Controllers/SprintsController.cs (hunk 6 above)
[the remark this answers is not in the document]
Comment (cristi-paraschiv) [reply IssueTrackerAPI/Controllers/SprintsController.cs:85]: nope, but they should be unassigned

[the remark this answers is not in the document]
Comment (GhiocelAndrei) [reply IssueTrackerAPI/Controllers/SprintsController.cs:85]: Added UnassignSprintFromIssuesAsync function in the latest commit.

File: IssueTrackerAPI/Controllers/SprintsController.cs (hunk 9 above)
[the remark this answers is not in the document]
Comment (GhiocelAndrei) [reply IssueTrackerAPI/Controllers/SprintsController.cs]: I see, renamed it in the latest commit.

File: IssueTrackerAPI/Controllers/IssuesController.cs [hunk 12]
```diff
@@ -53,8 +62,17 @@ public async Task<IActionResult> PutIssue(long id, IssueUpdatingDto issueDto, Ca
             return NoContent();
         }
 
+        // PUT: api/Issues/5
+        [HttpPut("Issues/AssignSprint")]
```
Comment (cristi-paraschiv) [inline IssueTrackerAPI/Controllers/IssuesController.cs:66]: why `Issues/`? how does the final endpoint look like to you?

File: IssueTrackerAPI/Controllers/IssuesController.cs [hunk 13]
```diff
@@ -40,8 +40,17 @@ public async Task<ActionResult<IssueDto>> GetIssue(long id, CancellationToken ct
             return _mapper.Map<IssueDto>(issue);
         }
 
+        // GET: api/Issues/Sprint/{id} 
+        [HttpGet("Issues/Sprint/{id}")]
```
Comment (cristi-paraschiv) [inline IssueTrackerAPI/Controllers/IssuesController.cs:44]: why `Issues/`? how does the final endpoint look like to you?

File: IssueTrackerAPI/Controllers/IssuesController.cs [hunk 14]
```diff
@@ -40,8 +40,17 @@ public async Task<ActionResult<IssueDto>> GetIssue(long id, CancellationToken ct
             return _mapper.Map<IssueDto>(issue);
         }
 
+        // GET: api/Issues/Sprint/{id}
```
Comment (cristi-paraschiv) [inline IssueTrackerAPI/Controllers/IssuesController.cs:43]: please stop adding this kind of comments, they don't bring anything useful; any developer should be able to infer the URL of the endpoint

File: IssueTrackerApplication/Services/SprintsService.cs [hunk 15]
```diff
@@ -0,0 +1,49 @@
...
+        }
+
+        public async Task CreateSprintWithIssuesAsync(CreateSprintWithIssuesCommand sprintDto, CancellationToken ct)
```
[the remark this answers is not in the document]
Comment (cristi-paraschiv) [reply IssueTrackerApplication/Services/SprintsService.cs:41]: seems like in-memory database does not support transactions

https://learn.microsoft.com/en-us/ef/core/testing/testing-without-the-database#transactions

File: IssueTrackerAPI/Controllers/SprintsController.cs [hunk 16]
```diff
@@ -0,0 +1,99 @@
...
+        private readonly IMapper _mapper;
+
+        public SprintsController(ISprintsService sprintsService, IMapper mapper)
+        {
+            _sprintsService = sprintsService;
+            _mapper = mapper;
+        }
+
+        // GET: api/Sprints
+        [HttpGet("Sprints")]
```
[the remark this answers is not in the document]
Comment (GhiocelAndrei) [reply IssueTrackerAPI/Controllers/SprintsController.cs]: Renamed it to 'All' also added 'Id' property in all DTOs that get returned in GET request

File: IssueTrackerAPI/Controllers/IssuesController.cs (hunk 12 above)
Comment (GhiocelAndrei) [reply IssueTrackerAPI/Controllers/IssuesController.cs]: Renamed it in the latest commit.

File: IssueTrackerAPI/Controllers/IssuesController.cs (hunk 13 above)
Comment (GhiocelAndrei) [reply IssueTrackerAPI/Controllers/IssuesController.cs]: Renamed it in the latest commit.

File: IssueTrackerAPI/Controllers/IssuesController.cs (hunk 14 above)
Comment (GhiocelAndrei) [reply IssueTrackerAPI/Controllers/IssuesController.cs]: Deleted unnecessary comments.

File: IssueTrackerApplication/Validations/SprintUpdatingValidator.cs [hunk 17]
```diff
@@ -0,0 +1,29 @@
...
+                .NotNull()
+                .WithMessage("Active flag is required.");
+
+            RuleFor(x => x.StartDate)
+                .NotEmpty()
+                .WithMessage("Start date is required.");
+
+            RuleFor(x => x.EndDate)
+                .NotEmpty()
+                .WithMessage("End date is required.");
```
[the remark this answers is not in the document]
Comment (GhiocelAndrei) [reply IssueTrackerApplication/Validations/SprintUpdatingValidator.cs:26]: This will be solved in Issue #31

**Comment by cristi-paraschiv** [comment]
please squash commits

**Pull request merged**